File: python/amostraMedia/todos.py
import random
import time

def heapify(arr, n, i):
    largest = i
    left = 2 * i + 1
    right = 2 * i + 2

    if left < n and arr[left] > arr[largest]:
        largest = left
    if right < n and arr[right] > arr[largest]:
        largest = right
    if largest != i:
        arr[i], arr[largest] = arr[largest], arr[i]
        heapify(arr, n, largest)

def heap_sort(arr):
    n = len(arr)
    for i in range(n // 2 - 1, -1, -1):
        heapify(arr, n, i)
    for i in range(n - 1, 0, -1):
        arr[i], arr[0] = arr[0], arr[i]
        heapify(arr, i, 0)

n = 100_000
print(f"\nExecutando Heap Sort com {n} elementos...\n")

arr_random = list(range(1, n + 1))
random.shuffle(arr_random)

print(" Caso ALEATORIO")
inicio = time.time()
heap_sort(arr_random)
fim = time.time()
tempo_aleatorio = fim - inicio
print(f"Tempo total: {tempo_aleatorio:.4f} s")
print("-" * 60)


arr_melhor = list(range(1, n + 1))

print(" Melhor caso (ordenado crescente)")
inicio = time.time()
heap_sort(arr_melhor)
fim = time.time()
tempo_melhor = fim - inicio
print(f"Tempo total: {tempo_melhor:.4f} s")
print("-" * 60)

arr_pior = list(range(n, 0, -1))

print("Pior caso (ordenado decrescente)")
inicio = time.time()
heap_sort(arr_pior)
fim = time.time()
tempo_pior = fim - inicio
print(f"Tempo total: {tempo_pior:.4f} s")
print("-" * 60)

print("\n COMPARATIVO FINAL")
print(f"Caso Aleatorio : {tempo_aleatorio:.4f} s")
print(f"Melhor Caso    : {tempo_melhor:.4f} s")
print(f"Pior Caso      : {tempo_pior:.4f} s")

File: C/amostraPequena/todos.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

void heapify(int arr[], int n, int i) {
    int largest = i;
    int left = 2 * i + 1;
    int right = 2 * i + 2;

    if (left < n && arr[left] > arr[largest])
        largest = left;
    
    if (right < n && arr[right] > arr[largest])
        largest = right;
    
    if (largest != i) {
        int temp = arr[i];
        arr[i] = arr[largest];
        arr[largest] = temp;
        heapify(arr, n, largest);
    }
}

void heapSort(int arr[], int n) {
    for (int i = n / 2 - 1; i >= 0; i--)
        heapify(arr, n, i);

    for (int i = n - 1; i > 0; i--) {
        int temp = arr[0];
        arr[0] = arr[i];
        arr[i] = temp;
        heapify(arr, i, 0);
    }
}

void preencherAleatorio(int arr[], int n) {
    for (int i = 0; i < n; i++)
        arr[i] = i + 1;

    for (int i = n - 1; i > 0; i--) {
        int j = rand() % (i + 1);
        int temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
    }
}

void preencherCrescente(int arr[], int n) {
    for (int i = 0; i < n; i++)
        arr[i] = i + 1;
}

void preencherDecrescente(int arr[], int n) {
    for (int i = 0; i < n; i++)
        arr[i] = n - i;
}

int estaOrdenado(int arr[], int n) {
    for (int i = 1; i < n; i++) {
        if (arr[i - 1] > arr[i])
            return 0;
    }
    return 1;
}

int main() {
    int n = 10000;
    int *arr = (int *)malloc(n * sizeof(int));
    if (!arr) {
        printf("Falha ao alocar memória.\n");
        return 1;
    }

    srand(time(NULL));

    printf("Executando Heap Sort com %d elementos...\n\n", n);

    preencherAleatorio(arr, n);
    printf("Caso ALEATÓRIO:\n");
    clock_t start = clock();
    heapSort(arr, n);
    clock_t end = clock();
    double tempoAleatorio = (double)(end - start) / CLOCKS_PER_SEC;
    printf("---------------------------------------------\n");

    preencherCrescente(arr, n);
    printf("Melhor caso (ordenado crescente):\n");
    start = clock();
    heapSort(arr, n);
    end = clock();
    double tempoMelhor = (double)(end - start) / CLOCKS_PER_SEC;
    printf("---------------------------------------------\n");

    preencherDecrescente(arr, n);
    printf("Pior caso (ordenado decrescente):\n");
    start = clock();
    heapSort(arr, n);
    end = clock();
    double tempoPior = (double)(end - start) / CLOCKS_PER_SEC;
    printf("---------------------------------------------\n");

    printf("\nCOMPARATIVO FINAL:\n");
    printf("Caso Aleatório : %.4f s\n", tempoAleatorio);
    printf("Melhor Caso    : %.4f s\n", tempoMelhor);
    printf("Pior Caso      : %.4f s\n", tempoPior);

    FILE *arquivo_saida;


    arquivo_saida = fopen("resultado.txt", "a");

    if (arquivo_saida == NULL) {
        printf("Erro ao abrir o arquivo resultado.txt!\n");
        free(arr);
        return 1; 
    }

    fprintf(arquivo_saida, "%.4f,%.4f,%.4f\n", 
            tempoAleatorio, 
            tempoMelhor, 
            tempoPior);

    fclose(arquivo_saida);

    printf("\nResultados salvos com sucesso em: resultado.txt\n");
    free(arr);
    return 0;
}
